core/src/apps/ethereum/sign_tx.py

```diff
@@ -99,16 +105,94 @@ async def sign_tx(
     return result
 
 
-async def handle_erc20(
+async def confirm_tx_data(
+    msg: MsgInSignTx,
+    defs: Definitions,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+    data_total_len: int,
+) -> None:
+    # function distinguishes between staking / smart contracts / regular transactions
+    from .layout import require_confirm_other_data, require_confirm_tx
+
+    if await handle_staking(msg, defs.network, to_address, maximum_fee, fee_items):
+        return
+
+    # Handle ERC-20, currently only 'transfer' function
+    token, recipient, value = await handle_erc20_transfer(msg, defs, to_address)
+
+    if token is None and data_total_len > 0:
+        await require_confirm_other_data(msg.data_initial_chunk, data_total_len)
+
+    await require_confirm_tx(
+        recipient,
+        value,
+        maximum_fee,
+        fee_items,
+        defs.network,
+        token,
+        bool(msg.chunkify),
+    )
+
+
+async def handle_staking(
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> bool:
+
+    data_reader = BufferReader(msg.data_initial_chunk)
+    if data_reader.remaining_count() < constants.SC_FUNC_SIG_BYTES:
+        return False
+
+    func_sig = data_reader.read_memoryview(constants.SC_FUNC_SIG_BYTES)
+    if msg.to in constants.ADDRESSES_POOL:
+        if func_sig == constants.SC_FUNC_SIG_STAKE:
+            await _handle_staking_tx_stake(
+                data_reader, msg, network, to_address, maximum_fee, fee_items
+            )
+        elif func_sig == constants.SC_FUNC_SIG_UNSTAKE:
+            await _handle_staking_tx_unstake(
+                data_reader, msg, network, to_address, maximum_fee, fee_items
+            )
+        else:
+            # unsupported function signature - will fall to blind signing
+            return False
+    elif msg.to in constants.ADDRESSES_ACCOUNTING:
+        if func_sig == constants.SC_FUNC_SIG_CLAIM:
+            await _handle_staking_tx_claim(
+                data_reader,
+                to_address,
+                maximum_fee,
+                fee_items,
+                network,
+                bool(msg.chunkify),
+            )
+        else:
+            # unsupported function signature - will fall to blind signing
+            return False
+    else:
+        # address does not correspond to staking smart contracts
+        return False
+
+    # successfully confirmed staking data
+    return True
```

Comment:
we prefer early returns to chaining if/elifs, as it makes it easier to be sure that some case doesn't fall through
```
if some_cond:
   if sig == xyz:
       await do_handle_xyz()
       return True
  if sig == abc:
       await do_handle_abc()
       return true

if some_other_cond:
    ...

# single final top-level catch-all
# no early-return happened so this was definitely not handled
return False
```

Reply:
Changed https://github.com/trezor/trezor-firmware/pull/3534/commits/339da2725da19a08e116a30fa035fcb17f31a463